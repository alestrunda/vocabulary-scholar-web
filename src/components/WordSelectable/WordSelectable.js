import React from 'react';
import PropTypes from 'prop-types';
import Checkbox from 'material-ui/Checkbox';
import classNames from 'classnames';
import { capitalizeFirstLetter, getAverageFromArray } from '../../misc';
import RatingStars from '../RatingStars';

class WordSelectable extends React.Component {
  state = {
    checked: false,
  };

  componentDidMount() {
    this.setState({
      checked: this.props.checked,
    });
  }

  handleCheck = () => {
    this.setState(
      {
        checked: !this.state.checked,
      },
      () => {
        this.props.onSelected &&
          this.props.onSelected(this.props.id, this.state.checked);
      }
    );
  };

  render() {
    return (
      <div
        className={classNames(
          'word-preview word-selectable',
          this.props.className
        )}
        onClick={this.handleCheck}
        style={this.props.style}
      >
        <div className="word-selectable__check">
          <Checkbox checked={this.state.checked} />
        </div>
        <div className="word-preview__inner word-preview__inner--left-big">
          <div className="el-full">
            <h3 className="word-preview__title">
              {capitalizeFirstLetter(this.props.word)}
            </h3>
          </div>
        </div>
        {this.props.ratings && (
          <div className="word-selectable__rating">
            <RatingStars stars={getAverageFromArray(this.props.ratings)} />
          </div>
        )}
      </div>
    );
  }
}

WordSelectable.propTypes = {
  className: PropTypes.string,
  id: PropTypes.string.isRequired,
  word: PropTypes.string.isRequired,
  ratings: PropTypes.array,
  onSelected: PropTypes.func,
  checked: PropTypes.bool,
  style: PropTypes.object,
};

WordSelectable.defaultProps = {
  checked: false,
};

export default WordSelectable;
