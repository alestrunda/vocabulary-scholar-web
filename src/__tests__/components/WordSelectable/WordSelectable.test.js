import React from 'react';
import { shallow } from 'enzyme';

import WordSelectable from '../../../components/WordSelectable';

it('renders correctly', () => {
  const props = {
    id: 'school',
    word: 'school',
  };
  const root = shallow(<WordSelectable {...props} />);
  expect(root).toMatchSnapshot();
});

it('triggers onSelected', () => {
  const props = {
    id: 'school',
    word: 'school',
    checked: true,
  };
  const onSelected = jest.fn();
  const root = shallow(<WordSelectable onSelected={onSelected} {...props} />);
  root.simulate('click');
  expect(onSelected).toHaveBeenCalledWith(props.id, !props.checked);
});
