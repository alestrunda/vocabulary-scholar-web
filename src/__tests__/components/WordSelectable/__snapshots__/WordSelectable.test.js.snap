// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WordSelectable
    checked={false}
    id="school"
    word="school"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="word-selectable__check"
        >
          <Checkbox
            checked={false}
            disabled={false}
            labelPosition="right"
          />
        </div>,
        <div
          className="word-preview__inner word-preview__inner--left-big"
        >
          <div
            className="el-full"
          >
            <h3
              className="word-preview__title"
            >
              School
            </h3>
          </div>
        </div>,
        undefined,
      ],
      "className": "word-preview word-selectable",
      "onClick": [Function],
      "style": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Checkbox
            checked={false}
            disabled={false}
            labelPosition="right"
          />,
          "className": "word-selectable__check",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "checked": false,
            "disabled": false,
            "labelPosition": "right",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="el-full"
          >
            <h3
              className="word-preview__title"
            >
              School
            </h3>
          </div>,
          "className": "word-preview__inner word-preview__inner--left-big",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h3
              className="word-preview__title"
            >
              School
            </h3>,
            "className": "el-full",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "School",
              "className": "word-preview__title",
            },
            "ref": null,
            "rendered": "School",
            "type": "h3",
          },
          "type": "div",
        },
        "type": "div",
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="word-selectable__check"
          >
            <Checkbox
              checked={false}
              disabled={false}
              labelPosition="right"
            />
          </div>,
          <div
            className="word-preview__inner word-preview__inner--left-big"
          >
            <div
              className="el-full"
            >
              <h3
                className="word-preview__title"
              >
                School
              </h3>
            </div>
          </div>,
          undefined,
        ],
        "className": "word-preview word-selectable",
        "onClick": [Function],
        "style": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Checkbox
              checked={false}
              disabled={false}
              labelPosition="right"
            />,
            "className": "word-selectable__check",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "checked": false,
              "disabled": false,
              "labelPosition": "right",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="el-full"
            >
              <h3
                className="word-preview__title"
              >
                School
              </h3>
            </div>,
            "className": "word-preview__inner word-preview__inner--left-big",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h3
                className="word-preview__title"
              >
                School
              </h3>,
              "className": "el-full",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "School",
                "className": "word-preview__title",
              },
              "ref": null,
              "rendered": "School",
              "type": "h3",
            },
            "type": "div",
          },
          "type": "div",
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
